File: exercise/week03/php/src/Game/RockPaperScissors.php
<?php

namespace Game;

class RockPaperScissors
{
    private const RULES = [
        '🪨' => [
            '✂️' => 'rock crushes scissors',
            '🦎' => 'rock crushes lizard',
        ],
        '📄' => [
            '🪨' => 'paper covers rock',
            '🖖' => 'paper disproves spock',
        ],
        '✂️' => [
            '📄' => 'scissors cuts paper',
            '🦎' => 'scissors decapitates lizard',
        ],
        '🦎' => [
            '🖖' => 'lizard poisons spock',
            '📄' => 'lizard eats paper',
        ],
        '🖖' => [
            '✂️' => 'spock smashes scissors',
            '🪨' => 'spock vaporizes rock',
        ],
    ];

    public static function play(string $player1, string $player2): array
    {
        if ($player1 === $player2) {
            return ['winner' => 'Draw', 'reason' => 'same choice'];
        }

        if (isset(self::RULES[$player1][$player2])) {
            return ['winner' => 'Player 1', 'reason' => self::RULES[$player1][$player2]];
        }

        return ['winner' => 'Player 2', 'reason' => self::RULES[$player2][$player1]];
    }
}

File: exercise/week03/java/src/main/java/game/RockPaperScissors.java
package game;

import java.util.EnumMap;
import java.util.Map;

public class RockPaperScissors {
    private static final Map<Choice, Map<Choice, String>> RULES = new EnumMap<>(Choice.class);

    static {
        RULES.put(Choice.ROCK, Map.of(
                Choice.SCISSORS, "rock crushes scissors",
                Choice.LIZARD, "rock crushes lizard"
        ));
        RULES.put(Choice.PAPER, Map.of(
                Choice.ROCK, "paper covers rock",
                Choice.SPOCK, "paper disproves spock"
        ));
        RULES.put(Choice.SCISSORS, Map.of(
                Choice.PAPER, "scissors cuts paper",
                Choice.LIZARD, "scissors decapitates lizard"
        ));
        RULES.put(Choice.LIZARD, Map.of(
                Choice.SPOCK, "lizard poisons spock",
                Choice.PAPER, "lizard eats paper"
        ));
        RULES.put(Choice.SPOCK, Map.of(
                Choice.SCISSORS, "spock smashes scissors",
                Choice.ROCK, "spock vaporizes rock"
        ));
    }

    private RockPaperScissors() {
    }

    public static Result play(Choice player1, Choice player2) {
        if (player1 == player2) return new Result(Winner.DRAW, "same choice");

        String reason = RULES.get(player1).get(player2);
        if (reason != null) return new Result(Winner.PLAYER_1, reason);

        return new Result(Winner.PLAYER_2, RULES.get(player2).get(player1));
    }
}